### ISSUE TITLE
History counters no longer are visible at breakpoints

### ISSUE THREAD

**Issue by pglvdm** [overturetool/overture#543]
This used to be the case and this information is essential when debugging concurrent VDM++ models. The smallest standard model where this is easy to see is the ReaderWriterPP example. Put a brakpoint in either the Write or Read operation and you will see it is not visible.

**Comment by nickbattle** [comment]
This processing is performed in the DBGPReaderV2 class, in the getContextValues method. If the code can determine that the current breakpoint is within an object context (ie. we can have history counters), and the class has a sync section (ditto), and the current Context is suspended on a sync guard, or the breakpoint line is the same as the sync definition or the lines within the sync clause, then it cycles through the sync clause creating name/value pairs for the history counters referenced and their current values. These name/value pairs are dumped into the environment returned, along with regular context variables, so they (should!) appear on the UI in the Variables view.

While debugging the Factorial spec, I am seeing history counters returned when a thread is blocked on a guard (eg. when another thread stops at a breakpoint). But I cannot get breakpoints on sync guards to trigger - they appear to be set, but they get ignored.

Can anyone else get breakpoints on sync clauses to stop?

**Comment by nickbattle** [comment]
While trying breakpoints, it looks like conditional and counter based breakpoints are not working either (or only intermittently)? The only breakpoints that I can get to work are "plain", without counters or conditions - these work fine as long as you're not in a sync clause.

**Comment by ldcouto** [comment]
I think at least some of this is an old issue: https://github.com/overturetool/overture/issues/362

**Comment by peterwvj** [comment]
I can't see history counters for the example below. Try to set a breakpoint inside the `SharedCounter.inc` operation.

I also can't get break points to work on guards.

~~~
class SharedCounter

instance variables

public counter : nat := 0;

operations

public inc : () ==> ()
inc () == counter := counter + 1;

public getValue : () ==> nat
getValue () == return counter;

sync
  per inc => #active(inc) = 0;

end SharedCounter

class Incrementer

instance variables

private sharedCounter : SharedCounter;

private isFinished : bool := false;

operations

public Incrementer : SharedCounter ==> Incrementer
Incrementer (s) == sharedCounter := s;

public joined : () ==> ()
joined () == skip;

public incCounter : () ==> ()
incCounter () ==
(
  for i = 1 to 100 do
    sharedCounter.inc();
  isFinished := true; 
)

thread
       incCounter();

sync
	per joined => isFinished;
			
end Incrementer

class Entry

operations

public static Run : () ==> ?
Run () ==
let s = new SharedCounter(),
    i1 = new Incrementer(s),
    i2 = new Incrementer(s),
    i3 = new Incrementer(s)
in 
(
  start(i1);
  start(i2);
  start(i3);

  i1.joined();
  i2.joined();
  i3.joined();
  
  return s.getValue(); 
);

end Entry
~~~

**Comment by pglvdm** [comment]
I am not certain that breakpoints in guards have worked at any point, but certainly I know that the possibility of inspecting the value of history counters when stopping at a breakpoint used to work (and is documented in the user manual).

**Comment by nickbattle** [comment]
This can be confusing. With the example you have, PJ, by stopping in inc() you have already passed inc's guard and therefore there are no history counters that matter for that thread at that moment. The main thread is blocked on the joined guard, but that sync expression does not include any #history counters in its expression - you only see the values of (say) #fin(op) if that is **used** in the guard that your thread is blocked on.

Can I ask you to try with the Factorial example, pasted below? Put a breakpoint on the "if" in doit, and ask for a modest factorial of say 10. After a few recursions, you should see that threads **other** than the one stopped at the breakpoint are stopped on giveResult, which includes #act(giveResult) and #fin(doit). Both of these therefore appear in the Variables view for the affected threads.

```
class Factorial
instance variables
	result : nat;

operations
	public Factorial: () ==> Factorial
	Factorial() == result := 0;
	
	public factorial : nat ==> nat
	factorial(n) ==
		if n = 0 then return 1
		else (
			dcl m : Multiplier := new Multiplier();
			m.calculate(1,n);
			start(m);
			result:= m.giveResult();
			return result
		)
		
end Factorial

class Multiplier
instance variables
	i : nat1 := 1;
	j : nat1 := 1;
	k : nat1 := 1;
	result : nat1 := 1
	
operations
	public calculate : nat1 * nat1 ==> ()
	calculate (first, last) ==
		(i := first; j := last);
		
	doit : () ==> ()
	doit() ==
	(
		if i = j then result := i   --- BREAKPOINT HERE
		else (
			dcl p : Multiplier := new Multiplier();
			dcl q : Multiplier := new Multiplier();
			start(p);
			start(q);
			
			k := (i + j) div 2;
			
			p.calculate(i,k);
			q.calculate(k+1,j);
			result := p.giveResult() * q.giveResult ()
		)
	);
	
	public giveResult : () ==> nat1
	giveResult() ==
		return result;
		
sync
	-- cyclic constraints allowing only the
	-- sequence calculate; doit; giveResult
	
	per doit => #fin (calculate) > #act(doit);
	per giveResult => #fin (doit) > #act (giveResult);
	per calculate => #fin (giveResult) = #act (calculate)
	
thread
	doit();
	
end Multiplier
```

**Comment by peterwvj** [comment]
![conc-fac](https://cloud.githubusercontent.com/assets/5388919/15215608/a82ddf2e-1853-11e6-8434-17b51b9e5ff3.png)

As shown in the picture I can inspect the history counters when a thread is stopped on `giveResult`.

But as far as I understand (and I'm a bit confused) PGL expects to see history counters when he is inside an operation? Can anyone get this to work?

Setting a break point on a sync guard does not work but I think that this is a known problem (see #362) as Luis pointed out. I've experienced problems with conditional breakpoints too.

**Comment by nickbattle** [comment]
But I'm not sure whether the history operator values are even defined when you are inside an operation. You cannot use the #operators other than in a sync clause. So what counters would be shown in an operation body? The counters for the operation itself or for all of the operations in the class or called operations or what? At least by showing **just** the history operations that are used in the sync clause, and **only** showing them when blocked in the sync clause, we are showing the values that are current and relevant.

Peter mentioned that something appears in the documentation. I don't have time to find it and check, but what is that showing?

**Comment by pglvdm** [comment]
ok I may have confused the issue but I was trying to find a way in which to see the value of the history counters (that I could see in the user manual) and since I couldn't set a breakpoint in the guard I couldn't see the history counters when I broke inside an operation controlled by a permission predicate. Thus, it just be the known problem mentioned by Luis. Btw the figure with the history counters are figure 6.4 on page 27.

**Comment by peterwvj** [comment]
![user-manual-debug](https://cloud.githubusercontent.com/assets/5388919/15216647/47bdcd0c-1858-11e6-994a-f5c546c7a41c.png)

Figure 6.4 seems to be similar to the situation shown in the figure in my previous comment.

**Comment by nickbattle** [comment]
Right. And if you look at that diagram, it seems to show a sync clause with a breakpoint, but it is definitely stopped (for whatever reason) **within** the sync clause, so the variables are showing.

**Comment by peterwvj** [comment]
Do we agree that this is not a bug then? If so we can close this issue.

**Comment by nickbattle** [comment]
I think so. We still have the one about not being able to set breakpoints in sync clauses though.

**Comment by peterwvj** [comment]
That's true but that's captured by issue #362

**Comment by nickbattle** [comment]
And does that also say that breakpoints with conditions and counters don't always work?

**Comment by peterwvj** [comment]
No you are right, but we should report it as a separate issue with a small example. I'm trying to write up a small example, but now it seems to work.. sigh :)

**Comment by peterwvj** [comment]
If you recall an example of a conditional/hit counter breakpoint that does not work please let me know.

**Comment by nickbattle** [comment]
I couldn't get either of them to work the other day - or rather, something worked once or twice but it was inconsistent. It was just a flat spec - shmem - BP somewhere in the main "count" loop.

**Comment by peterwvj** [comment]
I've been trying for around 20 minutes now and I can't produce the problem(s) using the shmem model. Could it be that the visual presentation of the breakpoints did not reflect the interpreter's view? Maybe the problem is something else. Hit counters seem to work for hits >= the user specified value. Although we expect that there exists a problem with conditional/hit counter break points I can't produce it. If we manage to produce the problem then I think we should report it in a separate issue.

**Comment by nickbattle** [comment]
OK, definitely happy to create a new bug as and when this appears.
